Source/FrameWork/GpuApi/GpuRhi.cpp

```diff
@@ -0,0 +1,209 @@
+#include "CommonHeader.h"
+
+#include "GpuRhi.h"
+
+#include "./Dx12/Dx12GpuRhiBackend.h"
+#include "GpuApiValidation.h"
+
+namespace FRAMEWORK
+{
+// Rhi validation layer, check validation before dispatch commands to concrete api backend.
+class GpuRhiValidation : public GpuRhi
+{
+public:
+	GpuRhiValidation(TUniquePtr<GpuRhi> InRhiBackend)
+		: RhiBackend(MoveTemp(InRhiBackend))
+	{
+	}
+
+public:
+	void InitApiEnv() override
+	{
+		RhiBackend->InitApiEnv();
+	}
+
+public:
+	void FlushGpu() override
+	{
+		RhiBackend->FlushGpu();
+	}
+
+	void BeginFrame() override
+	{
+		RhiBackend->BeginFrame();
+	}
+	void EndFrame() override
+	{
+		RhiBackend->EndFrame();
+	}
+
+	TRefCountPtr<GpuTexture> CreateTexture(const GpuTextureDesc &InTexDesc) override
+	{
+		return RhiBackend->CreateTexture(InTexDesc);
+	}
+
+	TRefCountPtr<GpuShader> CreateShaderFromSource(ShaderType InType, FString InSourceText, FString InShaderName, FString EntryPoint) override
+	{
+		return RhiBackend->CreateShaderFromSource(InType, InSourceText, InShaderName, EntryPoint);
+	}
+
+	TRefCountPtr<GpuShader> CreateShaderFromFile(FString FileName, ShaderType InType, FString EntryPoint, FString ExtraDeclaration) override
+	{
+		return RhiBackend->CreateShaderFromFile(FileName, InType, EntryPoint, ExtraDeclaration);
```

Comment:
```suggestion
		return RhiBackend->CreateShaderFromFile(MoveTemp(FileName), InType, MoveTemp(EntryPoint), MoveTemp(ExtraDeclaration));
```
Something like the above interfaces needs a `copy then move`  idiom to forward the parameters, which can reduce the cost of passing rvalue fstring.